keps/sig-api-machinery/1965-kube-apiserver-identity/README.md

```diff
@@ -95,20 +120,21 @@ will only delay the storage migration for the same period of time.
 
 ## Design Details
 
-The [kubelet heartbeat](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/0009-node-heartbeat.md)
+The [kubelet heartbeat](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/589-efficient-node-heartbeats)
 logic [already written](https://github.com/kubernetes/kubernetes/tree/master/pkg/kubelet/nodelease)
 will be re-used. The heartbeat controller will be added to kube-apiserver in a
 post-start hook.
 
-Each kube-apiserver will refresh its Lease every 10s by default. A GC controller
-will watch the Lease API using an informer, and periodically resync its local
-cache. On processing an item, the controller will delete the Lease if the last
-`renewTime` was more than `leaseDurationSeconds` ago (default to 1h). The
-default `leaseDurationSeconds` is chosen to be way longer than the default
+Each kube-apiserver will run a lease controller in a post-start-hook to refresh
+its Lease every 10s by default. A separate controller named [storageversiongc](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/storageversiongc/gc_controller.go)
+running in kube-controller-manager will watch the Lease API using an informer, and
+periodically resync its local cache. On processing an item, the `storageversiongc` controller
+will delete the Lease if the last `renewTime` was more than `leaseDurationSeconds` ago (default to 1h).
+The default `leaseDurationSeconds` is chosen to be way longer than the default
 refresh period, to tolerate clock skew and/or accidental refresh failure. The
 default resync period is 1h. By default, assuming negligible clock skew, a Lease
 will be deleted if the kube-apiserver fails to refresh its Lease for one to two
-hours. The GC controller will run in kube-controller-manager, to leverage leader
+hours. The `storageversiongc` controller will run in kube-controller-manager, to leverage leader
```

Comment:
I _think_ the main reason for running it in kube-controller-manager is to leverage leader election there. If we move this to kube-apiserver, wouldn't we need to wire in leader election there?

Reply:
Since this controller can work purely off of external observed state, it seems like a good candidate to run in KCM?

Reply:
It's fine if every apiserver runs the cleanup loop, for things that take an hour to expire it shouldn't hurt to race to delete them.

... I guess I don't care THAT much. Last argument: people forking & reusing apiserver (which I strongly discourage) but not controller manager will get a surprise

Reply:
> It's fine if every apiserver runs the cleanup loop, for things that take an hour to expire it shouldn't hurt to race to delete them.

I suppose this is true.  My natural inclination is still kube-controller-manager, but I probably won't fight too hard since active-active for this purpose is easy to reason about.